LoRaEngine/modules/LoRaWanNetworkSrvModule/Logger/IotHubLogger.cs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#nullable enable
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("LoRaWan.Tests.Unit")]
+[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2")]
+
+namespace Logger
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Text;
+    using System.Threading.Tasks;
+    using LoRaWan;
+    using Microsoft.Azure.Devices.Client;
+    using Microsoft.Extensions.DependencyInjection;
+    using Microsoft.Extensions.DependencyInjection.Extensions;
+    using Microsoft.Extensions.Logging;
+
+    public record IotHubLoggerConfiguration(LogLevel LogLevel, EventId EventId, bool UseScopes);
+
+    internal sealed class IotHubLoggerProvider : ILoggerProvider
+    {
+        private readonly ConcurrentDictionary<string, IotHubLogger> loggers = new();
+        private readonly Lazy<Task<ModuleClient>> moduleClientFactory;
+
+        public IotHubLoggerConfiguration Configuration { get; }
+        public LoggerExternalScopeProvider? ScopeProvider { get; }
+
+        public IotHubLoggerProvider(IotHubLoggerConfiguration configuration)
+            : this(configuration, new Lazy<Task<ModuleClient>>(ModuleClient.CreateFromEnvironmentAsync(new[] { new AmqpTransportSettings(TransportType.Amqp_Tcp_Only) })))
```

Comment:
Not sure why we need the lazy initialization here? Is it if the constructor gets called multiple times? That would be only in the case of a programming mistake because the DI should ensure the constructor is called only once, right?

Reply:
I think this is done to only create it, if we need to deliver a message. If the log level is not met, we will never create the instance - for example, if we only want to log errors, it won't get created unless we log an exception.

Reply:
I had two reasons in mind here: Since the module instantiation is asynchronous, I would need to take a `Task<T>` or `Func<Task<T>>` as a factory (at least), but I can't wait for the initialization to finish in the constructor. Since this is already additional complexity, I decided to use `Lazy<T>` instead, as it gives us the benefit of only executing when invoked.